clang/lib/Interpreter/CodeCompletion.cpp

```diff
@@ -56,26 +69,151 @@ class ReplCompletionConsumer : public CodeCompleteConsumer {
   std::shared_ptr<GlobalCodeCompletionAllocator> CCAllocator;
   CodeCompletionTUInfo CCTUInfo;
   std::vector<std::string> &Results;
+  ReplCodeCompleter &CC;
+};
+
+/*
+  The abstract class CompletionContextHandler contains four interfaces, each of
+  which handles one type of completion result.
+
+  Its substract classes are used to create concrete handlers based on
+  CodeCompletionContext.
+ */
+class CompletionContextHandler {
+protected:
+  CodeCompletionContext CCC;
+  std::vector<std::string> &Results;
+
+public:
+  CompletionContextHandler(CodeCompletionContext CCC,
+                           std::vector<std::string> &Results)
+      : CCC(CCC), Results(Results) {}
+
+  // convert a Declaration completion result  to a completion string, and then store it in Results.
+  virtual void handleDeclaration(const CodeCompletionResult &Result) {}
+
+  // convert a Keyword completion result  to a completion string, and then store it in Results.
+  virtual void handleKeyword(const CodeCompletionResult &Result) {}
+
+  // convert a Pattern completion result  to a completion string, and then store it in Results.
+  virtual void handlePattern(const CodeCompletionResult &Result) {}
+
+  // convert a Macro completion result  to a completion string, and then store it in Results.
+  virtual void handleMacro(const CodeCompletionResult &Result) {}
+};
+
+class DotMemberAccessHandler : public CompletionContextHandler {
+public:
+  DotMemberAccessHandler(CodeCompletionContext CCC,
+                         std::vector<std::string> &Results)
+      : CompletionContextHandler(CCC, Results) {}
+  void handleDeclaration(const CodeCompletionResult &Result) override {
+    auto *ID = Result.Declaration->getIdentifier();
+    if (!ID)
+      return;
+    if (!isa<CXXMethodDecl>(Result.Declaration))
+      return;
+    const auto *Fun = cast<CXXMethodDecl>(Result.Declaration);
+    if (Fun->getParent()->getCanonicalDecl() ==
+        CCC.getBaseType()->getAsCXXRecordDecl()->getCanonicalDecl()) {
+      LLVM_DEBUG(llvm::dbgs() << "[In HandleCodeCompleteDOT] Name : "
+                              << ID->getName() << "\n");
+      Results.push_back(ID->getName().str());
+    }
+  }
+};
+
+class DefaultAccessHandler : public CompletionContextHandler {
```

Comment:
What is the benefit of having a base class? We could merge the implementations of `DefaultAccessHandler` and `DotMemberAccessHandler` into the base class to simplify things.

Reply:
@vgvassilev Do you mean we don't need an "abstract" class? I am all for it. 

[I am a FP person. I just want to write high-order functions most of the time]